src/test/unit/BLSOperatorStateRetrieverUnit.t.sol

```diff
@@ -0,0 +1,239 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity =0.8.12;
+
+import "../utils/MockAVSDeployer.sol";
+
+contract BLSOperatorStateRetrieverUnitTests is MockAVSDeployer {
+    using BN254 for BN254.G1Point;
+
+    uint8 maxQuorumsToRegisterFor = 4;
+    uint256 maxOperatorsToRegister = 4;
+    uint32 registrationBlockNumber = 100;
+    uint32 blocksBetweenRegistrations = 10;
+
+    struct OperatorMetadata {
+        uint256 quorumBitmap;
+        address operator;
+        bytes32 operatorId;
+        BN254.G1Point pubkey;
+        uint96[] stakes; // in every quorum for simplicity
+    }
+
+    function setUp() virtual public {
+        _deployMockEigenLayerAndAVS();
+    }
+
+    function testGetOperatorState_Valid(uint256 pseudoRandomNumber) public {
+        (
+            OperatorMetadata[] memory operatorMetadatas,
+            uint256[][] memory expectedOperatorOverallIndices
+        ) = _registerRandomOperators(pseudoRandomNumber);
+
+        for (uint i = 0; i < operatorMetadatas.length; i++) {
+            uint32 blockNumber = uint32(registrationBlockNumber + blocksBetweenRegistrations * i);
+
+            (uint256 quorumBitmap, BLSOperatorStateRetriever.Operator[][] memory operators) = 
+                operatorStateRetriever.getOperatorState(registryCoordinator, operatorMetadatas[i].operatorId, blockNumber);
+
+            assertEq(operatorMetadatas[i].quorumBitmap, quorumBitmap);
+            bytes memory quorumNumbers = BitmapUtils.bitmapToBytesArray(quorumBitmap);
+            
+            _assertExpectedOperators(
+                quorumNumbers,
+                operators,
+                expectedOperatorOverallIndices,
+                operatorMetadatas
+            );
```

Comment:
trying to understand this -- for _each operator_, we are getting just that operator's bitmap, converting that bitmap to a compressed list of quorum numbers, and then feeding that list of quorum numbers into the `_assertExpectedOperators` function.
seems like we should probably be checking something smaller once per operator, or doing one big check for all operators.
I'm also a bit concerned on the structure here -- clearly different operators should have different `quorumNumbers`, so what this tells me is that if quorumNumbers is empty/lacking (for whatever reason) then the checks in `_assertExpectedOperators` should vacuously pass, which doesn't seem desirable.

Reply:
> as for speed, this file takes 6.37s for testing with 10 registering operators. tolerable imo

ah OK great, that's wayyy better than I feared. I agree that's acceptable.

Reply:
> in the assertion, we loop through each of the quorum numbers and check that the returned operator lists and states were as expected. we do this for every operator the check that this function works over time as more operators register. i could also just run it 2-3 times at different points in the list.
> 
> as for different operators having different quorumNumbers, that's correct, but we're specifically checking the functionality from the POV of the registering operator, who will only request for their quorums.

Ahhh, OK, gotcha, so the check is basically:
I register a new operator for `quorumNumbers`, then I want to check that all of `operators` have the correct `expectedOperatorOverallIndices` and `operatorMetadatas` (specifically, the `operatorMetadatas.stakes` are correct)

Makes sense, I think 👍 
We could in theory check that things remain the same for other quorum numbers (that the operator isn't registering for), but that would just mean way more "hollow" loop iterations, so I'm OK not doing that for now.
I'd appreciate if you can maybe help me break this down into properties that we can try to prove?

Reply:
I'd suggest adding something like my explanation to the documentation for the check here. IMO this PR is really great work, it's decently "self documenting", but it's hard to parse things when they're so complex -- a few brief, written explanations in normal language could help a lot!

Reply:
https://github.com/Layr-Labs/eigenlayer-contracts/pull/88/commits/7fae560531f2ad0ae1ff79289dd1c625b3368af5